### ISSUE TITLE
Inbox notification when downvoted question is edited

### ISSUE THREAD

**Issue by fixer1234** [soscripted/sox#98]
**Installed Version: 1.X.X  |  Environment: Chrome/Tampermonkey**
Note: Firefox not Chrome

I believe this is tied to the token issue (reported separately), but not 100% sure.  If so, this is already known and, hopefully, would disappear when the token problem is solved.

**Expected Behavior**

> How do you expect the feature to work?
> Feature button to select the option on the post, plus inbox notification when the downvoted question is edited.

**Actual Behavior**

> How is the feature currently behaving?

Feature button: This was originally an eyeball icon and my recollection is that the icon was replaced by a text link as one of the actions under the question.  That part worked in previous versions but is missing in v2.0.1 (perhaps now a function of the token?).

Notification: This part never worked for me.  I think this is the feature that the token is required for and that piece never worked, so the notification didn't.  The token is still not working (different issue already reported), so I assume that's what underlies the problem with this feature.

**Error Logs**

> SOX errors are logged to the browser console, if possible, please post those here.

I'll collect these if needed, but my understanding is that the problem is known (token).

**Steps to reproduce**

> What steps did you take to reproduce the issue?

Went to a question I had downvoted to test the feature, but was unable to test because of the missing button.

**Comment by shu8** [comment]
@fixer1234 the bug is fixed in the [new version](https://github.com/soscripted/sox/raw/fix/sox.user.js), so it should start working once the access token is sorted

**Comment by fixer1234** [comment]
@shu8,  I'll keep this feature issue separate from the token issue, even though they're related, in case any of the feature behavior is independent of the token.  

In the new version, the "notify on edit" link is back, but it isn't working correctly, possibly because of the token issue, which I'll update separately.  

It looks like you use a color change to indicate status on a post.  Clicking doesn't reliably cause the color change, it often requires several clicks.  

The change also isn't permanent.  If the page is reloaded, the color reverts to "inactive".  

If an edit has occurred and a message pops up to advise this, with instructions to click the message to refresh the post, the "notify on edit" link disappears with the refresh.  Reloading the page restores it (but without the active color indicator).

Aside from the button behavior, the feature does not result in notification, but my guess is that the token isn't working.

If all of the behavior is tied to the token, it probably makes sense to just wait on that issue.

**Comment by fixer1234** [comment]
I'll wait to retest this for the token fix.

**Comment by fixer1234** [comment]
@shu8, as mentioned in a comment on #71, I apparently have a token.  It isn't clear which token issue your comment above refers to, but:

The feature button appears to work in the new and improved, new v2.0.1.  The active indicator (color change), happens on the first click and persists through page reloads, etc.

The notification doesn't work yet.  I selected the notification on edit on three questions.  One was in the edit review queue for someone else's edit, which was approved.  The other two I subsequently edited.  By about half an hour later, there were no notifications in my inbox.

**Comment by shu8** [comment]
Should be fixed in [this even newer version](https://github.com/soscripted/sox/raw/fix/sox.user.js)

**Comment by fixer1234** [comment]
Newest, newer, new version:  I selected the Notify on Edit button on an answer post in the edit review queue.  A few minutes later, I got a notification that the post had been edited.  So that part works.  The only problem is that the edit hadn't been approved yet, so there was no edit on the post.

**Comment by fixer1234** [comment]
@shu8 The action on the page where the notification option is selected  seems to be working, and then all of the issues are behind the scenes in the site infrastructure.  The inbox isn't tied to any particular page, and the action is at unpredictable times.  Is there a practical way to capture logs of what's going on behind the scenes?

**Comment by shu8** [comment]
@fixer1234 This feature is intended for long-wait edits. It only checks for edits every day; the code looks at the date of the last checked time, and only requests updates from the API if the current date is after the last date.

eg. if I check an SE site today, date `26`, it won't look for any edits until at least the `27`th. Does that explain, a bit, of what's happening and why you're not getting updates when you think they should be?

I think the whole 'double-notification' thing is a bug, and as soon as I can repro it, I'll try to fix it.

To confirm the notification feature is working *as intended* though, please could you do the following:

- 'watch' a few posts
- edit them sometime on the *same* day
- check back the next day, and you should hopefully see a few notifications.

If you see double notifications, please do screenshot them and also dump the console contents here (the net tab is **not** required for this dump!)

**Comment by fixer1234** [comment]
@shu8 I jumped to v2.0.1f and will test as you suggest.  However, let me just say, "WHAT ON EARTH WERE YOU THINKING?"  :-)

There's no mechanism for being alerted to changes regardless of when they happen.  If the OP jumps right on a problem post and fixes it, we don't want to wait until the next day to find out about it.  The goal is to get the post back in circulation as fast as possible.  Why introduce a delay, even go out of your way to specifically code it?

Please, go into the code, grab that date test, and yank that sucker out.  Do it now.  Don't wait to pass go and collect $200.

This is one of SOX's most valuable features.  People have been asking for a way to be alerted to edits after close votes for years.  I even posted a question on conditional close votes a year ago for which this is the answer.  Now that we have a solution, don't crap it up.  If you build in a delay, people will still need another solution because this won't be it.

OK, I feel a little better now that I got that out of my system.

**Comment by shu8** [comment]
@fixer1234 Wow. OK. Sounds like that was a shit idea :P 

Any ideas as to what I can make the date test be then? Every (half an) hour? I just don't want to be polling the API every other second uselessly, because SE has some pretty strict throttle limits that block you for days if you do stuff like that.

I'd be more than happy to change it to anything else, provided it's not too often! :) Please do reply with timing thoughts :)

**Comment by fixer1234** [comment]
@shu8 So the plot thickens.  Too bad there isn't a way for the feature to leave a hook on the post that gets triggered when the post is edited.  The delay is something that it would probably be worth advising SOX users about so they don't think the feature isn't working.  

Timing thoughts?  Real time isn't necessary, but as close to it as practical.  Most users aren't glued to the site.  They periodically check it, or leave it open and glance for indications of activity.  The sooner the alert is posted, the sooner they can act; or the notification might catch therm on an earlier visit.  

I would think something on the order of 15 minutes would be great.  If that's too often, even hourly would work, especially if users were aware of the polling schedule.  If that's still too much polling, go with whatever you can get away with.  Two hours isn't nearly as good as one hour, but it's way better than next day.  Just my general sense that once you get to four hours or so, it might end up effectively next day for a lot of users, anyway (but if that's the best you can get away with, it's still better than next day for at least the diehard users).

If this becomes a popular feature (although SOX is one of the best kept secrets on SE), maybe the site programmers can build some kind of hook so SOX doesn't have to create so much traffic.  After all, people have been asking for some type of solution to close vote notification for a long time.

**Comment by shu8** [comment]
@fixer1234 I'll see if I can find something :)

**Comment by shu8** [comment]
@fixer1234 So, I've rewritten this feature, bearing in mind what you said.

The new version now does 2 things:

- uses the WebSockets to add notifications **live**, as they happen, as long as you are on an SE site somewhere, and
- uses the API, and checks **at least every hour** for updates

Please could you install [v2.0.1h](https://github.com/soscripted/sox/raw/fix/sox.user.js) and test this?

I haven't tested it lots myself yet, so it may be buggy, but please let me know if anything happens that you think shouldn't.

You can test the live-part by 'watching' a question on 1 tab, and then open another on a different tab, edit the 'watched' question, and you should see a notification appear on your other tab.

You can test the hourly, API, part by watching a question, editing it, and then come back in an hour on any SE page and you should see a notification.

If there are any bugs with the notification itself, please screenshot the notification in the inbox dropdown dialog and attach it here, and dump the console as well :)

**Comment by fixer1234** [comment]
@shu8 Will do.  Just so I know what I'm looking for, what role does SOX play after the post is selected?  i.e., does the selection trigger a behind-the-scenes process that uses the SE infrastructure, so the notification will appear even on a computer without SOX active, or does SOX need to be active (and v2.0.1h?) to get the subsequent notification?

**Comment by shu8** [comment]
@fixer1234 Needs to be on the (same) computer, with SOX v2.0.1h or later. It does use the SE infrastructure, partially, but all the data the script uses is saved locally so you need to use the same script and computer :/

BTW -- I know the green font change isn't persistent over page loads, but it is saved. I'll fix that tomorrow. :)

**Comment by fixer1234** [comment]
@shu8 The difference in result (and coordination), between live and API isn't clear (other than the delay); how do they co-exist?  Does API always provide a (second) notification as long as the Notify on edit is still selected?  In what case would it provide a notification if live has already done it?  

Also, if someone edits a post themselves, they already know about it.  They could be interested in subsequent edits by others.  They could wait to select Notify on edit until they're done their own edit, so that's a mechanism to avoid notification confusion.  I suppose being notified of your own edit is a way to check the feature.  Does the feature currently make any distinction between your own edits and those of others?

Anyway, some test results:

- Question in edit review queue, where I improved the edit (1st test): notification of the edit (not clear whether this was for the original edit or mine), appeared maybe a minute later, and only on the SU page I was currently on.  Closing the notification popup window without clicking Mark as read, cleared the inbox icon's new message count.  Returning later, the notification was gone from the inbox.  I didn't deselect Notify on edit.  No further notification anywhere.

- Question only I edited (in the same tab as selecting Notify on edit): no notification after an hour+.

- Question only I edited (in a different tab from selecting Notify on edit): no notification after an hour+.

- Question edited only by someone else:  Noticed that the edit had been approved 8 minutes after approval.  No notification on any tab at that time, and nothing in the inbox (so it isn't clear whether there was no live notification, or there was a temporary notification.  1+ hour later, no notification on any tab.

One possibility: have you incorporated something to prevent multiple notifications that might be suppressing all notifications after the first?

**Comment by shu8** [comment]
@fixer1234 In [v2.0.1i](https://github.com/soscripted/sox/raw/fix/sox.user.js), I've made the green font change persistent over page loads.

- currently, if the live notification comes up, and you mark it as read, then in an hour, the api notification should also come back (this will be fixed soon)
- for the api notification to appear, you need to refresh the other tab after an hour.

I've just thought (as I was writing this!) of how I could make it check for updates on every page load, so please bare with me whilst I implement it, so I wouldn't worry too much about testing the *api* part right now, but rather the live part please! :)

**Comment by shu8** [comment]
@fixer1234 ignore the above message, I just pushed [v1.0.2j](https://github.com/soscripted/sox/raw/fix/sox.user.js). 

Now, you can get live notifications, and the API will be checked for every post, hourly, provided you're still on an SE site.

Clicking 'mark as read', currently, does not remove it from the watch list.. what are your thoughts on this issue? Should clicking it remove it from the watch list? I'm 50/50 atm, because on one hand, a user might still want to receive edit notifications, however on the other hand, the use is decreased a bit because if the first edit didn't do much, you might think whether subsequent ones would either.. any thoughts?

**Comment by fixer1234** [comment]
@shu8 I'll try to test this later.

Thoughts on the mechanics:  I'm not sure if the action is designed the way it is because of limitations, or because you want it to work that way on purpose.  What I would find most intuitive would be this:

- Select **Notify on edit**.  That would persist regardless of whether the tab remained open.  Ideally, regardless of whether the browser was open or the computer was on.  If SOX needs to be active, a new session later could restart the action.  Any built-in delays would consider true elapsed time.  So if you're checking for something every hour and it's been at least an hour since the previous session, it would immediately check the status.  The active/inactive font color would survive refreshes, new sessions, etc.

- **Posting notification**.  I'm still missing the logic for live vs. API.  Some action would send a notification to the inbox; ideally immediately, otherwise within a 15 minutes to 1 hour kind of time frame (the sooner the better).  There would be just one notification.  This could get complicated if there are multiple edits.  I haven't thought through whether one notification per post would be better, or a notification for each edit.  I suppose if you really wanted to get fancy, there could be a notification for a subsequent edit only if there was an indication that the user was aware of the prior edit notification (mark as read or visited the post after the prior notification).  That's probably overkill, though.  Safe solution: notification for each edit.

- **Preserving notification**.  The notification is the only link back to the post.  If it gets deleted, there's no way for the user to find the post, again.  I don't think this should be deleted at all.  Treat it like any other message in the inbox history.  This also simplifies action for closing the cycle.

- **Closing the cycle**.  If notification persists in the inbox, there's no real need for Mark as read.  That leaves toggling the Notification on edit on the post.  Leaving that manual provides the ability to send notifications of subsequent edits, or the user can decide they don't care about subsequent edits and toggle it off.  If they forget to toggle it off, they could get future "spam" posts.  OTOH, the purpose is to be notified of an edit.  Sending a notification could automatically toggle off further notification.  It would be up to the user to turn it back on if they want to be alerted to future edits.  Not sure which is better.  The safer approach is to leave it a manual toggle and let the user be responsible.  Again, if you wanted to get really fancy, you could build a preference selection into the SOX menu so people can decide which way they prefer to use it.

If this type of action is precluded by limitations you have to work with, let me know and I'll rethink this.

**Comment by shu8** [comment]
thanks for that

- the first bullet point is exactly what currently happens
- basically, every time a post is edited, a notification will popup on
another tab if it exists. If another tab doesn't exist, then the script
will wait for *at least an hour*, before it notifies you of the edit. so if
you open your browser again 2 hours later, it would notify you immediately.
Of course, in normal day-to-day usage, the 'another tab' is actually your
main tab, because you wouldn't be the person editing the post, it would be
somebody else
- fair enough. I'll make it so notifications are preserved, and I'll remove
the 'mark as read' toggle (makes it much easier for me tbh :P), and
whenever a post is unmarked for watching, any notifications related to that
will be deleted as well -- sound good?
- I'll make it so that a user must manually unwatch a question, and
notifications will keep on coming until they say not to. In a future
release, I may add a settings toggle to this feature, to make the user
choose whether it should stop watching immediately or not.

Does all of this make sense? and does that 3rd bullet sound good?

**Comment by fixer1234** [comment]
@shu8 OK 2nd bullet doesn't seem to work as intended, yet (as of the previous version I tested).  I'm still not following all of the logic on this piece of the action.  If a person is doing their own edit, why is there a need to notify them at all (other than to test the feature)?  If it does notify on their own edit, why the need to do it on a different tab, using different timing?  If you're able to trigger a live notification, why not do only that?  If you need to rely on the API, why 2 hours instead of half an hour (as long as half an hour doesn't trigger throttling)?  I think the main consideration with the length of the delay is the likelihood that the action will be delayed by a day because we miss the window of the user's visit(s) that day.

Third bullet, I'm not sure of the point.  If you're saying new notifications stop when the user deselects watching, I agree.  If you're saying delete old notifications from the inbox, then no.  Once an appropriate notification is sent to the inbox, it should never be deleted, regardless of whether the user later stops watching that post.  They might want to think about whether an edit fixes a problem and then come back later to make a decision on retracting a downvote or close vote.  Or they might not have time to review or improve on a lengthy edit and want to come back later.  Treat it like any other inbox message; leave it in the history.

Fourth bullet sounds good.

**Comment by fixer1234** [comment]
@shu8 Tested v2.0.1k on #95 and got throttled.  I'll wait for instructions before testing this feature, especially if you plan to modify it.

**Comment by fixer1234** [comment]
@shu8 Installed v2.0.1L and then did a several hour session.  No sign of throttling, so that's good.  I'm not sure what you may have changed in the design of this feature, so not sure what I'm looking for.  Testing included tracking several edits.  One was approved 10 minutes ago, another was my edit from something over an hour ago.  Neither produced a fast notification, so it isn't clear whether you retained the live notification.  No notification was received on either one, so it isn't clear whether it isn't working (I can look for a notification tomorrow), or the delay is set too long to get a notification in time for it to be useful.  I'll check it tomorrow.  If there were no notifications, how do I usefully dump console output since the action is behind the scenes at a delayed time?

**Comment by shu8** [comment]
@fixer1234 ok, just to try of everything is working, (might be a bit of a pain...):

Firstly, please 'watch' a question, and then go to the site's homepage and stay there. Then, open a **private/incognito** window, login to SE and edit that same question. As soon as you edit it, go back to the non-private window and you should see a notification. If not, please dump the entire console here.

---

Regardless of whether that works, please also:

Watch a question and *close everything SE-related*. Open a **private/incognito** window and edit that same question. Close the private window and open a normal window again (where SOX is enabled) and within 15 minutes a notification will appear. You **will** need to refresh every few minutes to trigger the checking for updates. If it doesn't work after 15 minutes, please dump the console here as well

---

Changes I made:

- notifications never get deleted
- you can *manually* delete a notification
- questions are **not** automatically unwatched -- you need to manually unwatch them
- every 15 minute check rather than hourly
- use access token (so not throttling occurs)

**Comment by fixer1234** [comment]
@shu8 First test: no immediate notification (~5 minutes).  I'll check again after 15 minutes.  Console output:

```
GET 
http://superuser.com/questions [HTTP/1.1 200 OK 90ms]
GET 
http://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js [HTTP/1.1 304 Not Modified 39ms]
GET 
http://cdn.sstatic.net/Js/stub.en.js [HTTP/1.1 304 Not Modified 26ms]
GET 
http://cdn.sstatic.net/Sites/superuser/all.css [HTTP/1.1 304 Not Modified 33ms]
Warning: NetUtil.asyncFetch() requires the channel to have one of the security flags set in the loadinfo (see nsILoadInfo). Please create channel using NetUtil.newChannel()
GET 
https://i.stack.imgur.com/DPd3M.jpg [HTTP/1.1 304 Not Modified 28ms]
GET 
https://i.stack.imgur.com/DPd3M.jpg [HTTP/1.1 304 Not Modified 28ms]
GET 
https://www.gravatar.com/avatar/08537988ca97ed7fee568c386705566d [HTTP/2.0 304 Not Modified 21ms]
GET 
https://i.stack.imgur.com/lAhj3.png [HTTP/1.1 304 Not Modified 28ms]
GET 
https://www.gravatar.com/avatar/1d437dbb5f5309d4742523c79d9199c8 [HTTP/2.0 304 Not Modified 25ms]
GET 
https://i.stack.imgur.com/D0jX2.jpg [HTTP/1.1 304 Not Modified 31ms]
GET 
https://www.gravatar.com/avatar/a007be5a61f6aa8f3e85ae2fc18dd66e [HTTP/2.0 304 Not Modified 27ms]
GET 
https://www.gravatar.com/avatar/8cb7561f4a73c8ec8c2999c8927f7696 [HTTP/2.0 304 Not Modified 27ms]
GET 
https://i.stack.imgur.com/VI9St.png [HTTP/1.1 304 Not Modified 46ms]
GET 
https://i.stack.imgur.com/WfFBz.png [HTTP/1.1 304 Not Modified 37ms]
GET 
https://i.stack.imgur.com/Daf6l.jpg [HTTP/1.1 304 Not Modified 41ms]
GET 
https://www.gravatar.com/avatar/2c492983af89b49d44250c79579a2641 [HTTP/2.0 304 Not Modified 36ms]
GET 
https://www.gravatar.com/avatar/0b51840eb7be0a43d1d855b2f79bc0c8 [HTTP/2.0 304 Not Modified 40ms]
GET 
https://i.stack.imgur.com/ee4vJ.jpg [HTTP/1.1 304 Not Modified 44ms]
GET 
https://i.stack.imgur.com/i1gXV.jpg [HTTP/1.1 304 Not Modified 53ms]
GET 
https://www.gravatar.com/avatar/25521d528798772f5301d7454331cedb [HTTP/2.0 304 Not Modified 53ms]
GET 
https://www.gravatar.com/avatar/c5ef481e47f40861b734abc7d7973777 [HTTP/2.0 304 Not Modified 52ms]
GET 
https://www.gravatar.com/avatar/210849701af7002ec1074c3d2413749f [HTTP/2.0 304 Not Modified 53ms]
GET 
https://i.stack.imgur.com/yxD51.png [HTTP/1.1 304 Not Modified 59ms]
GET 
https://www.gravatar.com/avatar/0d1e9ff0945bdf0c66b9c33610bf2961 [HTTP/2.0 304 Not Modified 52ms]
GET 
https://i.stack.imgur.com/pX0Ho.jpg [HTTP/1.1 304 Not Modified 64ms]
GET 
https://i.stack.imgur.com/z21Fe.jpg [HTTP/1.1 304 Not Modified 63ms]
GET 
https://i.stack.imgur.com/a2FJQ.png [HTTP/1.1 304 Not Modified 70ms]
GET 
https://www.gravatar.com/avatar/c2b1989b07fa0729b669780ff3651b04 [HTTP/2.0 304 Not Modified 80ms]
GET 
https://www.gravatar.com/avatar/921c0a985c2c2ae7f64a76d17bf691d5 [HTTP/2.0 304 Not Modified 79ms]
GET 
https://www.gravatar.com/avatar/19623717c990a24ef0ae6486c0bfbef3 [HTTP/2.0 304 Not Modified 79ms]
GET 
https://www.gravatar.com/avatar/78dff053ebc39d2b62dd4aa32c256607 [HTTP/2.0 304 Not Modified 79ms]
GET 
https://www.gravatar.com/avatar/3f36e96fe935e4c0537ecdc527a43b25 [HTTP/2.0 304 Not Modified 78ms]
GET 
https://lh5.googleusercontent.com/-332g73eD8y8/AAAAAAAAAAI/AAAAAAAAAMs/R4ynYOn1jrc/photo.jpg [HTTP/2.0 304 Not Modified 71ms]
GET 
https://www.gravatar.com/avatar/a8a7ad36cd2dacfb50986e019d150bb8 [HTTP/2.0 304 Not Modified 78ms]
GET 
https://i.stack.imgur.com/W4Ip2.png [HTTP/1.1 304 Not Modified 74ms]
GET 
https://www.gravatar.com/avatar/339251ce3e514b07c355587ef216b95f [HTTP/2.0 304 Not Modified 77ms]
GET 
https://www.gravatar.com/avatar/646a1d7964f88fd9296f93e96c787c0a [HTTP/2.0 304 Not Modified 76ms]
GET 
https://www.gravatar.com/avatar/64aa7d1e0eb006175565adb0a47a4452 [HTTP/2.0 304 Not Modified 75ms]
GET 
https://i.stack.imgur.com/IVS5v.gif [HTTP/1.1 304 Not Modified 77ms]
GET 
https://lh6.googleusercontent.com/-TCfll-xz2KU/AAAAAAAAAAI/AAAAAAAAAH0/24G2UE1QnqU/photo.jpg [HTTP/2.0 304 Not Modified 75ms]
GET 
https://i.stack.imgur.com/ZMHzx.jpg [HTTP/1.1 304 Not Modified 83ms]
GET 
http://static.adzerk.net/ados.js [HTTP/1.1 304 Not Modified 95ms]
GET 
https://www.google-analytics.com/analytics.js [HTTP/2.0 304 Not Modified 82ms]
GET 
http://edge.quantserve.com/quant.js [HTTP/1.1 304 Not Modified 108ms]
GET 
http://cdn.sstatic.net/img/share-sprite-new.svg [HTTP/1.1 304 Not Modified 19ms]
GET 
http://cdn.sstatic.net/Sites/superuser/img/sprites.svg [HTTP/1.1 304 Not Modified 18ms]
GET 
http://cdn.sstatic.net/img/favicons-sprite16.png [HTTP/1.1 304 Not Modified 20ms]
GET 
https://i.stack.imgur.com/PVqCK.jpg [HTTP/1.1 304 Not Modified 30ms]
GET 
http://b.scorecardresearch.com/beacon.js [HTTP/1.1 200 OK 26ms]
XrayWrapper denied access to property "ready" (reason: value is callable). See https://developer.mozilla.org/en-US/docs/Xray_vision for more information. Note that only the first denied property access from a given global object will be reported.sox.common.js:18
GET 
http://cdn.sstatic.net/Js/full.en.js [HTTP/1.1 304 Not Modified 30ms]
SOX:  initializing SOX dialogsox.common.js:63:17
GET 
http://b.scorecardresearch.com/b [HTTP/1.1 204 No Content 20ms]
GET 
https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css [HTTP/2.0 304 Not Modified 16ms]
SOX:  injecting features into dialogsox.common.js:63:17
observe: .inbox-dialogsox.common.js:84:13
observe: .commentsox.common.js:84:13
SOX Access Token: SETsox.common.js:46:13
observe: #wmd-redo-buttonsox.common.js:84:13
observe: [id^="wmd-redo-button"]sox.common.js:84:13
observe: .image-upload formsox.common.js:84:13
Object { 142480: Object, 1120001: Object, 1120118: Object, 1120497: Object }sox.features.js:1468:13
Object {  }sox.features.js:1469:13
0sox.features.js:1470:13
Last Checked Time: 1472876518373sox.features.js:1555:21
SOX error: There was an error loading the feature "Unsorted-downvotedPostsEditAlert". Please raise an issue on GitHub, and copy the following error log:
SyntaxError: JSON.parse: unexpected character at line 1 column 2 of the JSON datasox.user.js:129:25

observe: .share-tipsox.common.js:84:13
GET 
https://www.google-analytics.com/collect [HTTP/2.0 200 OK 25ms]
GET 
http://pixel.quantserve.com/pixel;r=114655967;a=p-c1rF4kxgLUzNc;fpan=0;fpa=P0-1541000686-1462331864564;ns=0;ce=1;cm=;je=1;sr=1787x1005x24;enc=n;dst=1;et=1472876730767;tzo=240;ref=http%3A%2F%2Fsuperuser.com%2Fquestions;url=http%3A%2F%2Fsuperuser.com%2Fquestions;ogl=type.website%2Cimage.http%3A%2F%2Fcdn%252Esstatic%252Enet%2FSites%2Fsuperuser%2Fimg%2Fapple-touch-icon%402%252Epng%3Fv%3De869e4459439%2Ctitle.Recently%20Active%20Questions%2Cdescription.Q%26A%20for%20computer%20enthusiasts%20and%20power%20users%2Curl.http%3A%2F%2Fsuperuser%252Ecom%2Fquestions [HTTP/1.1 200 OK 53ms]
GET 
http://engine.adzerk.net/ados [HTTP/1.1 200 OK 46ms]
GET 
https://rawgit.com/shu8/SE-Answers_scripts/master/dupeClosedMigratedCSS.css [HTTP/2.0 304 Not Modified 35ms]
GET 
XHR 
https://query.yahooapis.com/v1/public/yql [HTTP/1.1 200 OK 111ms]
GET 
XHR 
https://api.stackexchange.com/2.2/questions [HTTP/1.1 200 OK 39ms]
GET 
XHR 
https://api.stackexchange.com/2.2/users/364367;242350;637025;270774;341338;510662;-1;286116;373245;486881;441207;136726;167207;21679;517391;102080;419383;637106;542362;277295;606191;559952;560638;234802;20006;637096;195972;94133;636634;73534;121352;637084;99868;17296;637082;388906;317323 [HTTP/1.1 200 OK 56ms]
GET 
XHR 
http://superuser.com/questions/1120034 [HTTP/1.1 301 Moved Permanently 45ms]
GET 
XHR 
https://api.stackexchange.com/users/ [HTTP/1.1 200 OK 66ms]
Object { items: Array[30], has_more: true, quota_max: 300, quota_remaining: 250 }sox.features.js:1823:17
GET 
http://static.adzerk.net/Extensions/adFeedback.js [HTTP/1.1 304 Not Modified 27ms]
GET 
http://static.adzerk.net/Extensions/adFeedback.css [HTTP/1.1 304 Not Modified 31ms]
GET 
http://qa.sockets.stackexchange.com/ [HTTP/1.1 101 Switching Protocols 50ms]
GET 
http://engine.adzerk.net/i.gif [HTTP/1.1 200 OK 50ms]
GET 
http://area51.stackexchange.com/ads/proposal/js [HTTP/1.1 200 OK 41ms]
GET 
XHR 
http://superuser.com/questions/1120034/configuring-multipleaccess-points-with-same-ssid [HTTP/1.1 200 OK 71ms]
GET 
https://qa.sockets.stackexchange.com/ [HTTP/1.1 101 Switching Protocols 78ms]
GET 
http://cdn.sstatic.net/area51/Img/bg-area51-ad-commitment.png [HTTP/1.1 304 Not Modified 20ms]
Use of getPreventDefault() is deprecated.  Use defaultPrevented instead.jquery.min.js:3:6598
```

I noticed a message in the middle that there was an error loading this feature.  It says to copy an error log.  I don't know if that is the log, or there is something somewhere else you need.

After coming here to post this, additional content showed up in the console.  Most is about github.  A small portion looks like it could potentially be relevant:

```
GET 
XHR 
https://services.addons.mozilla.org/en-US/firefox/api/1.5/search/guid:%7B972ce4c6-7e08-4474-a285-3208198ce6fd%7D,%7Be4a8a97b-f2ed-450b-b12d-ee082ba24781%7D,e10srollout%40mozilla.org,firefox%40getpocket.com,loop%40mozilla.org [HTTP/1.1 200 OK 433ms]
POST 
http://ocsp.digicert.com/ [HTTP/1.1 200 OK 78ms]
GET 
XHR 
https://versioncheck-bg.addons.mozilla.org/update/VersionCheck.php [HTTP/1.1 200 OK 379ms]
GET 
XHR 
https://versioncheck-bg.addons.mozilla.org/update/VersionCheck.php [HTTP/1.1 200 OK 485ms]
GET 
XHR 
https://versioncheck-bg.addons.mozilla.org/update/VersionCheck.php [HTTP/1.1 200 OK 298ms]
GET 
XHR 
https://versioncheck-bg.addons.mozilla.org/update/VersionCheck.php [HTTP/1.1 200 OK 586ms]
GET 
XHR 
https://versioncheck-bg.addons.mozilla.org/update/VersionCheck.php [HTTP/1.1 200 OK 207ms]
GET 
XHR 
https://rawgit.com/soscripted/sox/master/sox.user.js [HTTP/2.0 304 Not Modified 600ms]
POST 
http://ocsp.digicert.com/ [HTTP/1.1 200 OK 35ms]
1472876862948	addons.update-checker	WARN	Update manifest for {972ce4c6-7e08-4474-a285-3208198ce6fd} did not contain an updates property
1472876863138	addons.update-checker	WARN	Update manifest for firefox@getpocket.com did not contain an updates property
1472876863231	addons.update-checker	WARN	Update manifest for e10srollout@mozilla.org did not contain an updates property
GET 
XHR 
https://versioncheck-bg.addons.mozilla.org/update/VersionCheck.php [HTTP/1.1 200 OK 102ms]
GET 
XHR 
https://aus5.mozilla.org/update/3/SystemAddons/48.0.2/20160823121617/WINNT_x86-msvc-x64/en-US/release/Windows_NT%206.1.1.0%20(x64)/default/default/update.xml [HTTP/1.1 200 OK 343ms]
GET 
http://superuser.com/questions/1120497/how-to-colorize-cells-background-using-values-from-another-cells-in-openoffice [HTTP/1.1 200 OK 122ms]
When cancelling a request for http://cdn.sstatic.net/Sites/superuser/img/favicon.ico?v=29b26120d2c5 because the inner window was destroyed or a new favicon was loaded for it, it was already canceled!
PlacesUIUtils.jsm:109
GET 
http://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js [HTTP/1.1 304 Not Modified 56ms]
GET 
http://cdn.sstatic.net/Js/stub.en.js [HTTP/1.1 304 Not Modified 40ms]
GET 
http://cdn.sstatic.net/Sites/superuser/all.css [HTTP/1.1 304 Not Modified 46ms]
GET 
https://i.stack.imgur.com/DPd3M.jpg [HTTP/1.1 200 OK 38ms]
Warning: NetUtil.asyncFetch() requires the channel to have one of the security flags set in the loadinfo (see nsILoadInfo). Please create channel using NetUtil.newChannel()
GET 
https://i.stack.imgur.com/DPd3M.jpg [HTTP/1.1 304 Not Modified 23ms]
GET 
https://www.gravatar.com/avatar/e62d8d65ed2aa7b86bdc9890ef57a0da [HTTP/2.0 304 Not Modified 20ms]
GET 
http://cdn.sstatic.net/img/share-sprite-new.svg [HTTP/1.1 304 Not Modified 19ms]
GET 
http://cdn.sstatic.net/Sites/superuser/img/sprites.svg [HTTP/1.1 304 Not Modified 22ms]
GET 
http://cdn.sstatic.net/img/favicons-sprite16.png [HTTP/1.1 304 Not Modified 17ms]
The resource at "http://static.adzerk.net/ados.js" was blocked because tracking protection is enabled.[Learn More]how-to-colorize-cells-background-using-values-from-another-cells-in-openoffice
The resource at "https://www.google-analytics.com/analytics.js" was blocked because tracking protection is enabled.[Learn More]how-to-colorize-cells-background-using-values-from-another-cells-in-openoffice
The resource at "http://edge.quantserve.com/quant.js" was blocked because tracking protection is enabled.[Learn More]how-to-colorize-cells-background-using-values-from-another-cells-in-openoffice
The resource at "http://b.scorecardresearch.com/beacon.js" was blocked because tracking protection is enabled.[Learn More]how-to-colorize-cells-background-using-values-from-another-cells-in-openoffice
XrayWrapper denied access to property "ready" (reason: value is callable). See https://developer.mozilla.org/en-US/docs/Xray_vision for more information. Note that only the first denied property access from a given global object will be reported.sox.common.js:18
GET 
http://cdn.sstatic.net/Js/full.en.js [HTTP/1.1 304 Not Modified 21ms]
GET 
http://cdn.sstatic.net/Js/inline-tag-editing.en.js [HTTP/1.1 304 Not Modified 25ms]
SOX:  initializing SOX dialogsox.common.js:63:17
GET 
https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css [HTTP/2.0 304 Not Modified 19ms]
SOX:  injecting features into dialogsox.common.js:63:17
observe: .inbox-dialogsox.common.js:84:13
observe: .commentsox.common.js:84:13
SOX Access Token: SETsox.common.js:46:13
observe: #wmd-redo-button-1120497sox.common.js:84:13
observe: #wmd-redo-buttonsox.common.js:84:13
observe: [id^="wmd-redo-button"]sox.common.js:84:13
observe: .image-upload formsox.common.js:84:13
Object { 142480: Object, 1120001: Object, 1120118: Object, 1120497: Object }sox.features.js:1468:13
Object {  }sox.features.js:1469:13
0sox.features.js:1470:13
Last Checked Time: 1472876518373sox.features.js:1555:21
SOX error: There was an error loading the feature "Unsorted-downvotedPostsEditAlert". Please raise an issue on GitHub, and copy the following error log:
SyntaxError: JSON.parse: unexpected character at line 1 column 2 of the JSON datasox.user.js:129:25

observe: .share-tipsox.common.js:84:13
GET 
XHR 
https://query.yahooapis.com/v1/public/yql [HTTP/1.1 200 OK 212ms]
GET 
XHR 
https://api.stackexchange.com/2.2/users/364367;278537 [HTTP/1.1 200 OK 85ms]
GET 
XHR 
https://api.stackexchange.com/2.2/questions [HTTP/1.1 200 OK 83ms]
GET 
https://rawgit.com/shu8/SE-Answers_scripts/master/dupeClosedMigratedCSS.css [HTTP/2.0 304 Not Modified 35ms]
GET 
XHR 
http://superuser.com/posts/1120497/revisions [HTTP/1.1 200 OK 59ms]
GET 
XHR 
https://api.stackexchange.com/users/278537 [HTTP/1.1 200 OK 73ms]
"SOX editReasonTooltip URL: http://superuser.com/posts/1120497/revisions"sox.features.js:887:21
Object { context: <div#question.question>, length: 1, 1 more… }sox.features.js:888:21
SOX editReasonTooltip text: edited tagssox.features.js:889:21
Object { items: Array[1], has_more: false, quota_max: 300, quota_remaining: 248 }sox.features.js:1823:17
GET 
http://cdn.sstatic.net/Js/post-validation.en.js [HTTP/1.1 304 Not Modified 15ms]
GET 
http://cdn.sstatic.net/Js/external-editor.en.js [HTTP/1.1 304 Not Modified 20ms]
GET 
XHR 
http://superuser.com/posts/1120497/ivc/2d96 [HTTP/1.1 204 No Content 32ms]
GET 
http://qa.sockets.stackexchange.com/ [HTTP/1.1 101 Switching Protocols 51ms]
no element found2d96:1:1
The object cannot be linked to the inspector without a toolbox
console-output.js:3266
GET 
http://cdn.sstatic.net/Js/wmd.en.js [HTTP/1.1 304 Not Modified 17ms]
GET 
https://qa.sockets.stackexchange.com/ [HTTP/1.1 101 Switching Protocols 73ms]
Use of getPreventDefault() is deprecated.  Use defaultPrevented instead.jquery.min.js:3:6598
fire: #wmd-redo-buttonsox.common.js:91:29
fire: [id^="wmd-redo-button"]sox.common.js:91:29
GET 
http://cdn.sstatic.net/Sites/superuser/img/wmd-buttons.svg [HTTP/1.1 304 Not Modified 16ms]
observe: .commentsox.common.js:84:13
fire: .commentsox.common.js:91:29
SyntaxError: JSON.parse: unexpected character at line 1 column 1 of the JSON datasox.features.js:1526:29
GET 
XHR 
https://blocklist.addons.mozilla.org/blocklist/3/%7Bec8030f7-c20a-464f-9b0e-13a3a9e97384%7D/48.0.2/Firefox/20160823121617/WINNT_x86-msvc/en-US/release/Windows_NT%206.1/default/default/5/7/1/ [HTTP/1.1 200 OK 427ms]
SyntaxError: JSON.parse: unexpected character at line 1 column 1 of the JSON datasox.features.js:1526:29
```
Since the feature wasn't loaded in SOX, I'll assume no test is going to work, so I'll hold off on Test 2.

Should I assume that the notification trigger happens only once, immediately after editing?  For example, the other day, I set several posts to watch, but no notification showed up.  It's now a day after the edits were performed.  If I was to leave them set to watch, would the original notifications get generated at some future time when this feature is working?

**Comment by shu8** [comment]
@fixer1234 Can you try clicking 'reset' in the SOX settings dialog and then doing all this again? You'll need to re-get your access token as well. I think it's erroring out because I made a change to the structure of the saved data, and its trying to read the old version which doesn't exists anymore...

**Comment by fixer1234** [comment]
@shu8 I uninstalled the previous version, restarted Firefox, then installed v2.0.1L.  Still think reset will help?

**Comment by shu8** [comment]
It will definitely make a difference, because I don't think uninstalling
the script removes the settings that the script saved, however I don't know
whether it will fix the problem :/

**Comment by fixer1234** [comment]
@shu8 I need to re-enter every setting, so it isn't saving much.  :-)  I'll give it a try.

**Comment by fixer1234** [comment]
@shu8 I'm not sure reset is working.  The only obvious action it triggered was to stick the word reset next to the search box at the top.  None of the selections were changed, no new action to get a token, etc.  Also tried reset followed by saving changes (no obvious effect).  The feature is still in colored text with the pop-up tool tip that there was an error loading it.

**Comment by shu8** [comment]
There should be a reset button at the bottom of the dialog as well, next to
the save button

**Comment by fixer1234** [comment]
Yup, that's the one.

**Comment by fixer1234** [comment]
Should I post a bug report that the reset feature isn't working?  :-)

**Comment by enki-code** [comment]
@shu8 

Looks like there is some confusion here about how the settings are being stored. 

When installing a UserScript the script manager (Tampermonkey or Greasemonkey) assigns a GUID to the script, all settings for the script are stored with this GUID in an SQLite database.  So, a fresh install of the script should not be pulling old data from previous versions.

**Comment by shu8** [comment]
@enki-code hmm, ok...

BTW @enki-code and @fixer1234 everything over here's gone busy again with school starting again, so I might not be very active for the next few weeks :(

**Comment by fixer1234** [comment]
@shu8 Not clear what the status is on this one.  It got closed with a note on #110 that the bug should be fixed, but I noticed a milestone of v2.0.2.   Does that mean you don't expect this to work in this release?  

I loaded v2.0.1 and tested this.  No notification on four different questions after more than 3 hours.  I think this is where things left off earlier.  Should we restart diagnostics?  Would a new console dump be useful?  What procedure for testing as far as when to capture console contents?  

There doesn't appear to be a reopen choice here.  Should I post a fresh issue and start clean, given all the code changes you've made?

**Comment by shu8** [comment]
it shouldn't be working in this version... still a few more bugs to iron
out :( we'll need to restart diagnostics, but I don't have nearly as much
time that I can do that in. Might have to wait a few weeks, I'm really
sorry.

This wasn't meant to be closed, so I'll reopen it.

**Comment by shu8** [comment]
@fixer1234 just to update you on this... I've got the live part completely working (AFAIK) now, but the whole API-checking part is still broken. I'm trying to get this fixed ASAP, hopefully within the next 2 or so weeks! :)

sorry to keep you waiting!! 😞

**Comment by shu8** [comment]
@fixer1234 OK... this *should* be fixed in [the latest dev version e](https://github.com/soscripted/sox/raw/dev/sox.user.js).

Here's how it *should* work:

- You watch a post by clicking 'notify on edit' underneath it
- Now, if you are on any SE site, and someone edits a watched post, a WebSocket message is received and you should **instantly** get a notification that it's been edited; the notification should have `[LIVE]` appended on the end of it
- If you aren't on any SE site when a watched post it edited, then a notification will appear if the next time you go on an SE site is after 15 minutes of the edit.

Here's an example/scenario with the 2 cases:

- Live:
  - I watch a post on Super User
  - I start writing an answer to a different question (note: this is the 'still on an SE site part' I'm talking about)
  - Meanwhile, the post I watched has been edited
  - I get an instant notification

- Not live:
  - I watch a post on Super User
  - I go away and have dinner
  - I come back to any SE page in 13 minutes
  - Nothing happens
  - I come back 2 minutes later (15 mins altogether)
  - A notification is added when I go on any SE page

I hope this explains how it *should* work, and I hope it does work that way.

Please could you test if it works for you? :)

**Comment by fixer1234** [comment]
@shu8 Great news.  I'm worn out from installing several successive versions, but I'll load the new one when I get a second wind.  :-)

Question: it isn't clear why there are two notifications, and what happens to the live one.  If you can trigger an immediate notification, why not just make that the notification, and leave it there for whenever the user sees it?  

What happens to the live notification?  The description implies that it appears for only a limited time.  So say I'm in the middle of something and catch that an alert has been posted.  By the time I'm free to deal with it, might it be gone (and after close to 15 minutes it will show up again)?

On the Non-live notification, Is SOX collecting actions as they happen over a 15 minute period and waiting to display them, or every 15 minutes it checks for actions and reports anything that happened within the last 15 minute window, or the 15 minutes is some type of delay between each action and when that action is reported?

BTW, if you leave, have dinner, and get back on the computer in 13 minutes, you should take more time and savor the meal.  :-)

**Comment by shu8** [comment]
All notifications, no matter which type, remain in the inbox until they are
manually deleted.

There are 2 notifications because the live notification only works in
certain cases (when the edit happens when you are on a site) but the 15
minute one works even when a page isn't open.

Also, it checks every 15 minutes.

**Comment by fixer1234** [comment]
@shu8 Interesting.  So if an edit happens while the user isn't logged into an SE site, they find one notification.  If they're logged into a site when the edit happens, they get two?  A future feature might be to have any live notification suppress a second, non-live one.  Just an idea for another day.  Thanks.  I'll get on this and test later today.

**Comment by fixer1234** [comment]
@shu8 Initial tests:

- Live notification is successfully triggered, but action needs a little tweaking.

- Notification doesn't get posted to inbox until any site page is refreshed (not necessarily an issue, at least for me, because of frequent page refreshes).

- The notification has a delete button in place of the date/time.  Personally, I would view the notifications the same as any other, and have no intention or need to delete the history.  For me, the date/time would be more useful.  But that's a design decision.

- The message count is inaccurate.  On other notifications, the unread message count is zeroed as soon as you open the inbox, and then starts over with subsequent notifications.  These notifications don't get zeroed after viewing.  So subsequent notifications result in a cumulative count rather than a count of unread messages.

- Triggering of the notification makes the Notify on Edit button disappear if the page is still open.  The button reappears if the edited page is refreshed.  I didn't test whether this has any effect on notification of any subsequent edits.  Assuming visibility of the button doesn't affect its operation, the only drawback might be forgetting to toggle it off.

- Author names have disappeared from all posts in the inbox, which I assume is some kind of conflict between these two features.  I tested deselecting notifications to see if names would return to the inbox.  They didn't immediately return, although I didn't test waiting until the inbox was refreshed with a new message.  If the notification were to include the author name, the author would logically be the post editor.

Things I haven't tested yet:

- I generally have the site open somewhere, and leave the computer on around the clock most of the time, so it would likely be awhile before I could test the non-live notification.

- Haven't tested notification of subsequent edits to the same post, with or without reading the previous notification.  The expected behavior would be that subsequent edits would each generate a notification, although a refinement might be suppression of subsequent notifications if the notification for a previous edit has not yet been viewed.

- Haven't tested the feature with an answer post.

- Haven't test notifications if a question and an answer on the same thread are both being watched and both have edits.

- Haven't tested the effect of button visibility on notification of subsequent edits.

- Haven't tested eventual restoration of author names in inbox if notifications are deselected.

**Comment by shu8** [comment]
I think the easiest way to fix bullets 3, 4, 5, 6 and 8 would be to separate the inbox with this feature, and make a new dialog box solely for this feature's notifications, similar to the mod-diamond (new meta questions) feature and the SOX's own cogs dialog.

How does that sound? And also, do you have any ideas on what the button image could be for the dialog? (ie. instead of cogs/mod diamond)

---

I'll get the other bugs fixed soon :)

**Comment by fixer1234** [comment]
@shu8 Clever solution.  Icon?  What jumps to mind would be one of these (no particular order): 

![su pencil](https://cloud.githubusercontent.com/assets/10065766/19898322/fdb48734-a031-11e6-8c75-92b5f7257169.png)
![su edit](https://cloud.githubusercontent.com/assets/10065766/19898335/0cab7996-a032-11e6-9228-28692a820a2d.png)
![su bell](https://cloud.githubusercontent.com/assets/10065766/19898365/2396028e-a032-11e6-96ee-62357236cca7.png)
![su bell1](https://cloud.githubusercontent.com/assets/10065766/19898386/2eb68238-a032-11e6-8a89-dcd9eda2d343.png)
![su check](https://cloud.githubusercontent.com/assets/10065766/19898400/3bf72b78-a032-11e6-891e-9aa4ba3e515c.png)
![su check1](https://cloud.githubusercontent.com/assets/10065766/19898412/44b385e0-a032-11e6-9426-cdfbb1810ecc.png)
![su hand](https://cloud.githubusercontent.com/assets/10065766/19898423/4d3f24c6-a032-11e6-82c5-4fd9855ad548.png)
![su star](https://cloud.githubusercontent.com/assets/10065766/19898435/54b99e7a-a032-11e6-9330-e6c98827814d.png)
![su thumbtack](https://cloud.githubusercontent.com/assets/10065766/19898443/5c6ddc9e-a032-11e6-97dd-22fa4544c86b.png)
![su filetext](https://cloud.githubusercontent.com/assets/10065766/19898450/6538d824-a032-11e6-8d26-e3332dc0226f.png)
![su filetext1](https://cloud.githubusercontent.com/assets/10065766/19898460/6d3c6f68-a032-11e6-8489-90b8fe48a30f.png)

**Comment by shu8** [comment]
hmm I quite like the `edit` one... I think I'll go with that one :) I'll
hopefully get this done in the next few days

**Comment by shu8** [comment]
@fixer1234 this is done in the [latest dev version g](https://github.com/soscripted/sox/raw/dev/sox.user.js).

It now has its **own** dialog solely for notifications from this feature.

As you'll see, the notification itself is pretty empty right now. Any ideas on what extra useful information can be added?

---

Regarding bullet 2: that shouldn't happen. No refreshes should be needed, so please tell me if that happens again in the new version.

There's no notification number now; the button simply glows red if there are unread notifications.

Live and non-live notifications work in exactly the same way

**Comment by fixer1234** [comment]
@shu8 Just a preliminary check.  I watched a question, edited it, and later found the notification in the new box.  So the basic functionality is working.

I'm not sure when it showed up or whether page refresh triggered it showing up because I'm colorblind; if the icons is glowing red, I'd never know.  Bright blue, like you use for the Mod diamond would be more visible.

I'll do some testing and report back.  And I'll think about additional info that could be useful.  

A few initial thoughts.  The name of the editor would probably be useful (easy way to differentiate my own edits, edits by the core contingent of experienced users, and the OP).  Rough edit timeframe, like in the regular inbox, would be useful.  Another thought, the edit comment.  Another possibility, some type of indicator for the status of the watch button (maybe display the eye icon if watch is active).

**Comment by fixer1234** [comment]
@shu8 Very minor bug discovered with this feature.  The new message box is open when opening the SU chat room selection page (http://chat.stackexchange.com/?tab=site&host=superuser.com).  It covers parts of 4 chat rooms, so it has to be closed to use the page.  So far, that's the only page I've seen the symptom on.

![su chatrooms](https://cloud.githubusercontent.com/assets/10065766/20033980/36ce77f4-a385-11e6-96f1-6e31746c2011.png)

**Comment by shu8** [comment]
That's been fixed in the latest version (I think it's P now). You can
install it using any links I've ever given in these issues :) (I'm not on
my computer atm so can't give you the link!)

**Comment by shu8** [comment]
@fixer1234 I didn't think about the colour being a problem.. sorry! do you have any thoughts on what I could use instead of colour? because I realise that different people can't see different colours...

regarding the extra info, I'll add the:

- editor
- edit comment
- Q/A
- edit time

but I don't get what you mean about the indicator (your last bullet point)..?

---

I think adding separate notifications for each edit is going to get really complicated and would cause users who use the feature to just rely on the edit comment as a determining factor of whether they should go look at the post again or not. I think it's best to just have a notification if the post was *edited at all*, and if it is, then the user should go to the post and check it out. 

I do kind of see your point in seeing the advantage of seeing all the different edits, but I think that would just get too complicated TBH.

Also, what do you think of changing the link of the notification to the **revision page** rather than the question page? that would make users more likely to see the difference...

**Comment by shu8** [comment]
@fixer1234 I've added those into [dev z](https://github.com/soscripted/sox/raw/dev/sox.user.js) :)

Can you see how it works for you? :)

**Comment by fixer1234** [comment]
@shu8 I'll load dev z today and test.  re: bullet points and questions -- 

- status indicator referred to whether the Notify on edit selection is still active on the post (way to see that in the message window without needing to go back to the post).  As an aside, that's another case where the color change isn't very visible if you're color blind, mainly because the font used as the button is thin and there isn't a lot of color in one spot.  It would be more visible if the button text went to bold in addition to the color change, or even change the button to display the eyeball icon instead of the Notify on edit text when it is active.  My bullet suggested adding the eyeball icon to the notification in the message box if the feature is enabled on the associated post.

- re: edit comment, not sure I understand this one.  Here's what I was referring to.  SOX has a working feature to display the edit comment on hover, which it captures when the page is loaded.  I was referring to making this part of the message content, similar to how an excerpt of a normal comment is used as the "body" of the message in the regular message inbox.  Regular comments are stored and indexed by SE, so they can be directly retrieved.  I'm assuming that for watched posts that are edited and therefore generate a message, SOX would need to capture the edit comment the same way it does for the hover feature, and then store this itself for use in the message window.

- re: color of the top bar icon, there are a number of potential ways to improve this.  Red-green color blindness is by far the most common, and even for non-color blind people, red is not the most visible color.  Green is the most visible color, but that's not easy for red-green color blind people to differentiate.  Low sensitivity to blue or yellow is pretty rare, so using pure blue or yellow for the active state vs. grey for the inactive state is better.

   The amount and concentration of color is also a big factor.  If the icon or text is composed of thin lines, so there isn't a lot of color in one spot, it's hard to see differences in color.  The mod diamond works because it's blue and is a substantial blob of color.  So using an icon with substantial uninterrupted mass, or switching to bold text for "active", makes the color more visible (in addition to a change in appearance).  Changing the background color of the icon would be another way to make it very visible.  Completely changing the symbol (e.g., Notify on edit text to an eyeball icon when active), is pretty foolproof.

- Re: separate notifications, I don't know the mechanics of how you generate or filter the notifications, so there isn't much I can comment.  I had envisioned just adding a notification to the message box any time a different author edited the post; the user would be responsible for deleting what they considered obsolete or redundant messages.  Here's the main issue.  The message is the link to the post for the user to be able to get back to it later, so users will keep the notification in the message box after looking at the edit.  If no subsequent edits are reflected in the notification window, the user will never know about new edits, so the feature would no longer serve its purpose.  So some mechanism is needed for the user to be aware of subsequent edits.  Some potential solutions:

1. If you want just one notification displayed per post, display only the newest notification.  Particularly if you are including the editor ID and edit time, that will be a pretty good indicator.

2. Add some type of indicator flag to the message that indicates the existence of one or more edits after the last time the user either visited the post or looked in the message box (not sure if this is something capturable).

**Comment by fixer1234** [comment]
@shu8 Quick question.  You mentioned that this version includes some form of Q vs A identifier in the message box.  For all notifications, I'm seeing a cool icon, thread title, editor, edit date/time, and a delete button, but no obvious Q vs. A identifier.  Am I missing something or is this a bug?  example:

![sox edited posts](https://cloud.githubusercontent.com/assets/10065766/20589169/c3afdc14-b1e8-11e6-9ca9-93f9b34bf37f.png)

I believe there were both question edits.  Just got my first notification of an answer edit and it looks the same.

**Comment by shu8** [comment]
@fixer1234 sorry, my bad - I forgot to actually add the q/a! fixed in [dev aa](https://github.com/soscripted/sox/raw/dev/sox.user.js)

I'll start adding your suggestions in starting from the next version :)

**Comment by shu8** [comment]
@fixer1234 How's it looking now in [dev ar](https://github.com/soscripted/sox/raw/dev/sox.user.js)? :)

I've...:

- changed the red topbar icon to blue
- change the topbar icon to get bolder when it is showing a new message has popped up
- added an 'eye' to the notification to show whether watching is still active
- made sure the 'Q'/'A' is shown
- fixed the links -- they were actually often wrong links(!) :/

---

The problem with the edit comment thing is that the feature you're talking about which shows the edit comment as a tooltip only works on the **current site** because it manually scrapes the page for the edit comment.

Now, I could technically scrape for the edit comment in this feature as well, **but**, that would only work for edits on the current site (ie. if a question on SU was edited but I was on SO, then Javascript won't let me access another domain 😞) and it would be really inconsistent between sites then :/

**Comment by fixer1234** [comment]
@shu8  Good stuff.  I'll load it and check it out.

Interesting info on the edit comments.  A few thoughts:

- Many (most?) users predominantly use a single site.  For example, 99+% of my SE time is on SU; I venture to other sites for specific, unusual needs.  So a current site limitation might not be a limitation at all for many users.

- For users who do frequent a number of different sites, site-based comments could still be useful (when you go to a site, the comments for that site are visible).  Not quite a consolidated message box, but you need to go to the site to see the posts, anyway, so it would still be useful.

- I'm wondering whether there might be a workaround for sharing the comments across sites.  Some ideas: 

   * SOX parses links across sites (is that because the information is available via a different, live, mechanism that doesn't work for the comments?)

   * SOX stores it's own preferences and applies them across sites.  Could the relevant comments be scraped and stored centrally in a similar manner?

   * Could you use something like browser cookies to store the scraped comments and access them across sites?  Most people don't use multiple browsers concurrently, and this would limit the access to the same computer, but isn't that already the case for this feature, anyway?

**Comment by shu8** [comment]
@fixer1234 never mind, I think I've got a nice workaround... I'll update you after I've tried it :)

**Comment by shu8** [comment]
@fixer1234 Turns out I was completely wrong! There *is* [a revisions API method](https://api.stackexchange.com/docs/revisions-by-ids)! I have no idea how I ever missed that!

But the good news is; I've added the revision comment to the notification now, from [dev ax](https://github.com/soscripted/sox/raw/dev/sox.user.js) onwards! 😃 

Just hover over the question title in the notification and the tooltip will be the revision comment :)

I'm going to close this issue now, because it's no longer *broken*, but feel free to comment back if anything's still missing! :)

**Comment by fixer1234** [comment]
@shu8 Haven't done exhaustive testing yet, but this is shaping up to be a very complete and polished mini-app.  Most of the features work great.  A few minor glitches:

- The topbar icon doesn't reflect anything.  To me, all of the SE and SOX icons look grey on a black topbar in their "inactive" state.  The moderator diamond turns visibly blue if there are new Meta questions.  This message box icon doesn't get bold or change to blue when there are messages.  So that feature doesn't appear to be working.

- This may just be a limitation, but a number of elements don't update until the page is refreshed.  I'll have to review the presence or deletion of the messages, themselves (I made a note that refresh is required for the message box to update, and now I don't recall the specific details).  But the eyeball indicator I know is not updated until a refresh.  On the post, itself, clicking the Notify on edit button changes color immediately, but the bold text (a really nice addition, BTW), doesn't appear until a refresh.  

   Needing a page refresh isn't a big deal if people are aware of it, but if message box changes could be independent of the page, the status shown would be more "reliably accurate" (like the way the new message counter and the post ages on the main page update without refreshing the page).  

- Notification is never immediate.  Enabling Notify on edit and then doing an edit, doesn't produce a message immediately or within a few minutes.  But it does show up in a reasonable time.

- One action I haven't had a chance to test yet, I'm wondering how it is intended to work:  Say I enable Notify on edit, perform an edit, and then unable Notify on edit before the message appears.  Does the edit immediately register and it is just the populating of the message box that's periodic (so it will still show up), or is the presence of an edit scanned for periodically and the message  immediately populated, so if Notify on edit is not enabled at scan time, there will be no message?

- I edited a tag wiki and noticed that the Notify on edit button was there.  The situation was that I saw a proposed wiki edit in a review queue and wanted to propose another change after that review was complete.  I went to the tag wiki page and enabled Notify on edit to be alerted when the review was complete.  That edit was approved, but there was no notification, and the Notify on edit button was no longer enabled.  

   It would be handy if this worked, but I'm sort of not surprised that it didn't.  Wasn't sure if the wiki editor just shares some hooks so the button appears even though you don't expect it to work there, or it works only for edits by users with sufficient rep to not need review, or it should work there but doesn't.

**Comment by shu8** [comment]
@fixer1234 thanks! I'll look into the first 3 bullet points

- about your 4th one: as soon as you unwatch a post, the script won't check for edits on it at all again. (it checks every 15 minutes btw). so you were right in saying `if Notify on edit is not enabled at scan time, there will be no message`.

- about your 5th one: it's not *supposed* to work on tag wikis yet -- I'll try and disable the button on those pages. I just think the use case for tag wikis is much smaller than normal posts, and a bit more complicated because of the way they work. Either way, it's not on my todo list *at the moment*, at least! :)

**Comment by fixer1234** [comment]
@shu8 Now using Dev J and running into some buggy behavior with a number of aspects of this feature.  

First some misc. observations:

- When an edit is saved locally, or an external edit has occurred and you click the warning that there was an edit, click to refresh the page, the Notify on edit button disappears (regardless of whether active).  Refreshing the page restores it.  A direct tag edit (>10K rep feature) does not affect the button, but those actions are not fully reflected until a page refresh, anyway.

- The position of the Notify on edit button is not uniform.  It is to the right of the action buttons at the bottom of the post except for posts marked as a duplicate, in which case it's below the action buttons.

- The Notify on edit button immediately goes bold now, no page refresh required.

- I assume the new number of messages icon now replaces the bold/blue icon change when messages are present?

- All I ever saw was a (1).  Is that a fixed icon that represents any count (i.e., non-empty), or is that supposed to show a message count?

- If it is a count, does the count represent the total messages in the box or the number of unread messages (new since the box was last viewed)?

The rest of the input is a bit complex, with a lot of different behaviors, so I'll just relate the sequence of events:

- Notify on edit was active on two messages in the Edit review queue.  I'll refer to them as Post1 and Post2.  I left both tabs open.  Eventually, I returned to the Post1 tab and found that it had been approved; it had the warning message that the post had been edited, click to refresh.   

- Clicking that button showed the edited post and the topbar icon immediately changed to (1).  I didn't open the box yet because I wanted to see how the count would be affected by Post2.

- I clicked to some other open SU tabs.  The message box on those still showed the old icon, which didn't change if I refreshed the page.  I returned to the Post1 tab; it was the only one that showed the (1) icon.  It was the active tab when the icon changed.

- Still on Post1, and assuming the Post1 message was in the box, I deselected Notify on Edit.  The (1) icon immediately disappeared and it returned to the old icon.

- I went to the Post2 tab.  That one had also been approved and had the warning message about clicking to refresh to see the edit.  I clicked it, the post refreshed, registering the edit, but the topbar icon did not change to indicate the presence or increase of messages.

- At that point, I opened the message box on Post2.  There was no message about Post1, only a message for Post2.  No way to know whether there was ever a Post1 message, or some action caused it to be deleted.  

- I deselected Notify on edit on Post2, refreshed the page and the message was still there (with the eyeball now gone).  This would seem to point to some interaction with the new topbar icon, in which displaying that icon caused the message not to be populated.

- I had done some other watched edits, and checked periodically for a change in the topbar icon to indicate that the message had been produced.  There was never any icon change on any of those tabs.  So I started periodically checking the message box.  About 20 minutes after another edit, a second message appeared  in the box.

- When I left Post1, it was displaying the old icon.  After the activity described above, I returned to Post1, and it was displaying the (1) icon again (triggered by the edit that produced the second message?).  I refreshed the page, and the (1) icon disappeared; replaced with the old icon.  Both of the messages remained in the box.

- There were about half a dozen other watched edits.  At the time of writing this, it's going on 2 hours since those edits, but no others have produced a message.

There are so many different strange things on so many different pages, I'm not sure where to start in terms of capturing console output.  I figured some of this would probably be diagnostic of some programming oversights or common denominators, which might narrow the focus.

**Comment by fixer1234** [comment]
@shu8 I'll add to the saga:

- After completing the above post, I deselected Notify on edit on all of the open tabs and deleted the messages in the box.  There was one edited post still actively watched where I had accidentally closed the tab earlier, so I couldn't find it to deselect it.

- I enabled Notify on edit on several new edits.  Roughly 2 minutes after one of those edits, the (1) icon popped up on the page I happened to be on (review queue).  The message box contained two messages.  The lower one was from several hours ago, the "lost" actively watched edit.  The upper one was one of the new edits.

- The second new edit was not included.  Going to that post, I found the Notify on edit button deselected.

- Again, the message box was limited to 2 messages.

**Comment by shu8** [comment]
I just noticed something from one of your console dumps in another issue: the live feature doesn't work for meta posts right now. That might be the cause for some, but not all, of these issues?

**Comment by fixer1234** [comment]
@shu8 That might also be a problem, but none of what I tested involved Meta posts.

**Comment by fixer1234** [comment]
@shu8 I had to switch back to the 2.0.2 official version because of the FF hanging problem.  It has this feature in the menu, but no Notify on edit buttons on the posts.  It's hard to reconstruct the history, but it seems like at some point, the basic functionality was working (maybe late November-early December?).  Then there were a bunch of enhancements and basic changes and we ran into a new batch of problems.  So it isn't clear whether for v2.0.2, you rolled things back to the basic functionality or just partially disabled the feature.  Is your expectation that this feature should work in v2.0.2 (with no bells or whistles), or it's disabled and on hold for debugging?

**Comment by shu8** [comment]
@fixer1234 I'm not sure to be honest 😕 

I think it should have been working; I closed this 13 days ago and pushed to master 12 days ago... :/

Nevertheless, I'm working on your FF crash bug, your hint about it being to do with the comments seems right... working on it

**Comment by shu8** [comment]
Replacing this feature with #215. Closing.
